config/config_auth_details.go

```diff
@@ -0,0 +1,131 @@
+package config
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/slices"
+)
+
+// AuthDetails contains the details of the authentication configuration.
+type AuthDetails struct {
+	AuthType      string            `json:"auth_type"`
+	Host          string            `json:"host,omitempty"`
+	Configuration AuthConfiguration `json:"configuration"`
+}
+
+func (a *AuthDetails) String() string {
+	s := ""
+	if a.Host != "" {
+		s += fmt.Sprintf(`host=%s`, a.Host)
+	}
+
+	if a.AuthType != "" {
+		s += fmt.Sprintf(", auth_type=%s", a.AuthType)
+	}
+
+	conf := a.Configuration.String()
+	if conf != "" {
+		s += fmt.Sprintf(`
+Configuration:
+%s`, conf)
+	}
+
+	return s
+}
+
+// AuthConfiguration is a map of attribute name to its configuration.
+type AuthConfiguration map[string]*AttrConfig
+
+func (c AuthConfiguration) String() string {
+	var conf []string
+	for _, a := range ConfigAttributes {
+		if _, ok := c[a.Name]; ok {
+			k := a.Name
+			v := c[k]
+			conf = append(conf, fmt.Sprintf("- %s=%s", k, v))
+		}
+	}
+	return strings.Join(conf, "\n")
+}
+
+type AttrConfig struct {
+	Value            string `json:"value"`
+	Source           Source `json:"source"`
+	AuthTypeMismatch bool   `json:"auth_type_mismatch"`
+}
+
+func (a *AttrConfig) String() string {
+	s := fmt.Sprintf(`%s (from %s)`, a.Value, a.Source.String())
+	if a.AuthTypeMismatch {
+		s += " (not used by the current auth type)"
+	}
+	return s
+}
+
+type AuthDetailsOptions int
+
+const (
+	ShowSensitive AuthDetailsOptions = iota
+)
+
+func (c *Config) GetAuthDetails(opts ...AuthDetailsOptions) AuthDetails {
+	var showSensitive bool
+	for _, opt := range opts {
+		switch opt {
+		case ShowSensitive:
+			showSensitive = true
+		}
+	}
+
+	attrSet := make(map[string]*AttrConfig, 0)
+	for _, a := range ConfigAttributes {
+		if a.IsZero(c) {
+			continue
+		}
+
+		attrSet[a.Name] = &AttrConfig{
+			Value:            getValue(c, &a, showSensitive),
+			Source:           c.getSource(&a),
+			AuthTypeMismatch: a.HasAuthAttribute() && c.AuthType != "" && !slices.Contains(a.AuthTypes, c.AuthType),
+		}
+	}
+
+	return AuthDetails{
+		AuthType:      c.AuthType,
+		Host:          c.Host,
+		Configuration: attrSet,
+	}
+}
+
+func (c *Config) SetAttrSource(attr *ConfigAttribute, source Source) {
```

Comment:
Does this need to be public?

Reply:
@mgyucht yes, it's being used in CLI to set a different attribute source which does not exist in SDK (`bundle` to be specifc)